src/common/rpc/transaction.rs

```diff
@@ -139,102 +134,106 @@ type ReqId = Vec<u8>;
-    ) {
-        executor.spawn_ok(async move {
-            let mut msg_iterator = provider.iter();
-            while let Some(req) = msg_iterator.next() {
-                Self::submit_message(req, sender.clone());
-            }
-        });
-    }
-
-    async fn dispatch_loop(sender: Arc<SenderState>) {
-        while sender.is_open.load() {
+    async fn dispatch_loop(&self) {
+        while self.is_open.load() {
             const DISPATCH_INTERVAL: Duration = Duration::from_millis(3);
             sleep(DISPATCH_INTERVAL);
-            Self::dispatch_messages(sender.clone()).await;
+            self.dispatch_messages().await;
         }
     }
 
-    async fn dispatch_messages(sender: Arc<SenderState>) {
-        sender.ongoing_task_count.fetch_add(1);
-        let msgs = mem::take(&mut *sender.queued_messages.lock().unwrap());
-        if !msgs.is_empty() {
-            // let len = msgs.len();
-            sender.req_sink.clone().send(client_msg(msgs)).await.unwrap();
+    async fn dispatch_messages(&self) {
+        self.ongoing_task_count.fetch_add(1);
+        let messages = mem::take(&mut *self.queued_messages.lock().unwrap());
+        if !messages.is_empty() {
+            self.req_sink.clone().send(client_msg(messages)).await.unwrap();
         }
-        sender.ongoing_task_count.fetch_sub(1);
+        self.ongoing_task_count.fetch_sub(1);
     }
 
-    async fn await_close_signal(
-        close_signal_receiver: CloseSignalReceiver,
-        sender: Arc<SenderState>,
-    ) {
+    async fn await_close_signal(&self, close_signal_receiver: CloseSignalReceiver) {
         match close_signal_receiver.await {
             Ok(close_signal) => {
-                Self::close(sender, close_signal).await;
+                self.close(close_signal).await;
             }
             Err(err) => {
-                Self::close(sender, Some(Error::new(err.to_string()))).await;
+                self.close(Some(Error::new(err.to_string()))).await;
             }
         }
     }
 
-    async fn close(sender: Arc<SenderState>, error: Option<Error>) {
-        if let Ok(true) = sender.is_open.compare_exchange(true, false) {
-            if let None = error {
-                Self::dispatch_messages(sender.clone()).await;
+    async fn close(&self, error: Option<Error>) {
+        if let Ok(true) = self.is_open.compare_exchange(true, false) {
+            if error.is_none() {
+                self.dispatch_messages().await;
             }
             // TODO: refactor to non-busy wait?
             // TODO: this loop should have a timeout
             loop {
-                if sender.ongoing_task_count.load() == 0 {
-                    sender.req_sink.clone().close().await.unwrap();
+                if self.ongoing_task_count.load() == 0 {
+                    self.req_sink.clone().close().await.unwrap();
                     break;
                 }
             }
         }
     }
 }
 
-impl Drop for Sender {
-    fn drop(&mut self) {
```

Comment:
This `drop` stalled in a single-threaded environment and crashed the thread in a multi-threaded one. Project-wide correct `Drop` implementation is out of scope for this PR.